runtime/src/accounts_db.rs

```diff
@@ -1307,28 +1337,104 @@ impl AccountsDB {
                     trace!(
-                        "id: {} setting count: {} cur: {}",
+                        "slot: {} id: {} setting cur: {} full_count: {}",
+                        slot,
                         id,
-                        count,
-                        store.count_and_status.read().unwrap().0
+                        store.count_and_status.read().unwrap().0,
+                        full_count,
                     );
-                    store.count_and_status.write().unwrap().0 = *count;
+                    store.count_and_status.write().unwrap().0 = full_count;
                 } else {
-                    trace!("id: {} clearing count", id);
-                    store.count_and_status.write().unwrap().0 = 0;
+                    let mut c = store.count_and_status.write().unwrap();
+                    trace!(
+                        "slot: {}, id: {} resetting count old value: {:?}",
+                        slot,
+                        id,
+                        *c
+                    );
+                    c.0 = 0;
                 }
             }
+            if is_dead {
+                trace!("slot: {}: clearing dead slot", slot);
+                dead_slots.push(*slot);
+            } else {
+                trace!("slot: {}: NOT clearing dead slot", slot);
+                alive_slots.push(*slot);
+            }
+        }
+
+        let mut reclaims: Vec<(u64, AccountInfo)> = vec![];
+        accounts_index.clear();
+        for alive_slot in &alive_slots {
+            let mut accumulator: Vec<HashMap<Pubkey, (u64, AccountInfo)>> = self
+                .scan_account_storage(
+                    *alive_slot,
+                    |stored_account: &StoredAccount,
+                     store_id: AppendVecId,
+                     accum: &mut HashMap<Pubkey, (u64, AccountInfo)>| {
+                        let account_info = AccountInfo {
+                            store_id,
+                            offset: stored_account.offset,
+                            lamports: stored_account.account_meta.lamports,
+                        };
+                        debug!(
+                            "ryoqun: alive_slot: {}, pubkey: {:?}",
+                            *alive_slot,
+                            (stored_account.meta, &account_info)
+                        );
+                        accum.insert(
+                            stored_account.meta.pubkey,
+                            (stored_account.meta.write_version, account_info),
+                        );
+                    },
+                );
+
+            let mut account_maps = accumulator.pop().unwrap();
+            while let Some(maps) = accumulator.pop() {
+                AccountsDB::merge(&mut account_maps, &maps);
+            }
+            for (pubkey, (_, account_info)) in account_maps.iter() {
+                accounts_index.insert(*alive_slot, pubkey, account_info.clone(), &mut reclaims);
```

Comment:
We had to be conservative here; `purge_zero_lamport_account` didn't work since the introduction of #8218 after generate_index and before verify_snapshot().

Corresponding test  is this https://github.com/solana-labs/solana/pull/8337/files#diff-2099c5256db4eb5975c8834af38f6456R2307

Reply:
`convervative` means previously we only kept the last entry for each given pubkey in account index. That's very aggressive in a way. That's because [this](https://github.com/solana-labs/solana/pull/8337/files#diff-2099c5256db4eb5975c8834af38f6456R1415) is updating the index after marking the updated slot as `root` before that.